src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -3,136 +3,50 @@
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     internal static class CodeRefactoringHelpers
     {
-        public static Task<bool> RefactoringSelectionIsValidAsync(
-            Document document, TextSpan selection, SyntaxNode node, CancellationToken cancellation)
-        {
-            return RefactoringSelectionIsValidAsync(document, selection, node, ImmutableArray<SyntaxNode>.Empty, cancellation);
-        }
-
         /// <summary>
         /// <para>
-        /// Determines if a refactoring should be offered for a given node, given the specified selection in a document.
-        /// The refactoring is offered either if the selection is zero-width and not inside one of the specified holes
-        /// or if the selection contains the entire node and perhaps some whitespace.
+        /// Determines if a <paramref name="node"/> is underselected given <paramref name="selection"/>.
         /// </para>
         /// <para>
-        /// Note: this function considers the span containing the node to start at the earliest preceding whitespace
-        /// (including up to one newline) before the node’s <see cref="SyntaxNode.SpanStart"/> and up through the last
-        /// whitespace following the <see cref="SyntaxNode.Span"/>’s <see cref="TextSpan.End"/>.
+        /// Underselection is defined as ommiting whole nodes from either the beginning or the end. It can be used e.g. to detect that
+        /// following selection `1 + [|2 + 3|]` is underselecting the whole expression node tree.
         /// </para>
         /// <para>
-        /// For the <paramref name="holes"/>, the position is considered invalid if it is *within* the hole, not if it
-        /// is touching the hole edges.
+        /// It's intended to be used in conjunction with <see cref="IRefactoringHelpersService.GetRelevantNodesAsync{TSyntaxNode}(Document, TextSpan, CancellationToken)"/>
+        /// that, for non-empty selections, returns the smallest encompassing node. A node that can, for certain refactorings, be too large given user-selection even though
+        /// it is the smallest that can be retrieved.
         /// </para>
         /// </summary>
-        public static async Task<bool> RefactoringSelectionIsValidAsync(
-            Document document,
-            TextSpan selection,
-            SyntaxNode node,
-            ImmutableArray<SyntaxNode> holes,
-            CancellationToken cancellationToken)
+        public static bool IsNodeUnderselected(SyntaxNode node, TextSpan selection)
```

Comment:
>IsNodeUnderselected [](start = 27, length = 19)

Is this helper used only by the string interpolation refactoring? If so, consider moving it directly under the refactoring, unless you envision this helper will be helpful for other refactorings.

Reply:
Nope, it's used by IntroduceLocal as well. And potentially by other expression-based refactorings.